Infinite tree

kinow opened the pull request:
These changes close #330, and close #346, and close #256 (no recursion - no need for so many events, it's now controlled by the `infinite-tree` library), and close #227 (memory is about the same initially, but doesn't skyrocket over time), and close #222 

Now also closes #403 , as I used it to improve the JS code, and avoid having to build cycle points by looking at all the families (thanks to @dwsutherland 's tip).

Reverts #232 because the offline mode uses `cylc client graphql` to fetch data, and it doesn't get the `__typename`, which is set only by the JS ApolloClient.

**Requirements check-list**
- [X] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [X] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.


**Pending**

- [x] avoid data duplication
- [x] update the template to use the latest layout improvements #396 
- ~~try to add the Infinite Tree as a separate component, instead of replacing the current Tree, and use as an extra Lumino widget~~
- [x] check again that event listeners are being properly removed, widgets are being removed from DOM and telling the Vue components to be removed too, double check the scroll is working as expected, etc (there was a recent rebase)
- [x] add note about extra application size (~130 KB if I remember well, better double check later how much more we are adding here)
- [x] fix add view menu (I think my rebase/merge messed it and I lost graph/mutations)
- [x] keep user state, instead of overriding it with every websockets update (blocked by deltas https://github.com/cylc/cylc-flow/pull/3500 & https://github.com/cylc/cylc-uiserver/pull/118)

kinow commented:
Besides the pending items in the issue description, it is pending also:


- [ ] clean up commits
- [ ] fix unit test

kinow commented:
The main differences now from my previous tests (when I was using just `infinite-tree` and its render function) are:

- `infinite-tree` is not handling the DOM. It's only updating its data structures, and `vue-virtual-scroller` is doing the infinite scroll (this way we can use vue components)
- we now have `Task` and `Job` components in the DOM/vue (note to self, `vue-virtual-scroller` works better with **non-functional** components - see their `README`)
- most of the objects in memory were `Node` objects (from `infinite-tree`)
- I have duplicated memory because I am not using the vuex data structure (yet, will update it in the next days)
- most of the CPU/processing time is spent in scripting, which is converting the GraphQL data to the workflow tree data - this should get a lot better with incremental updates

Another thing that happened after this change, is that the `Node` has a `.state` attribute, that holds information like `depth`, `total` (how many children/grandchildren/etc, descendants?), and `open`.

So the expand/collapse state is in the data, not in the Vue component. So as we are re-creating the whole data structure with every data received from websocket, the `Node`s you expanded in the UI are then closed if that's their default state (e.g. a task is closed by default), or open (e.g. a cycle point is open by default).

kinow commented:
(Work-in-progress comment, will be updated over the next days)

## What improved

### DOM element hierarchy (See ref 1)

> Slightly trickier — reduce the size of your DOM tree and the number of elements in each branch. The smaller and shallower your document, the quicker it can be reflowed. It may be possible to remove unnecessary wrapper elements if you’re not supporting older browsers.

Before we had a recursive component, based on Vuetify tree, and on existing tutorials. While it worked, a change in the job details panel could trigger a reflow of its parent, and of its parents' parents, and so it goes. All the way up to the cycle point element, or the tree component element itself.

Now each entry in the tree is a node, with a `margin-left` proportional to its depth - the higher the depth, the wider the margin is. Updates to a job panel, now should happen within the job panel element. If a reflow occurs, it will be a bug, that can be fixed by changes in the CSS or HTML elements.

1 - https://www.sitepoint.com/10-ways-minimize-reflows-improve-performance/

kinow commented:
## What improved

### DOM element hierarchy (See ref 1)

> Slightly trickier — reduce the size of your DOM tree and the number of elements in each branch. The smaller and shallower your document, the quicker it can be reflowed. It may be possible to remove unnecessary wrapper elements if you’re not supporting older browsers.

Before we had a recursive component, based on Vuetify tree, and on existing tutorials. While it worked, a change in the job details panel could trigger a reflow of its parent, and of its parents' parents, and so it goes. All the way up to the cycle point element, or the tree component element itself.

Now each entry in the tree is a node, with a `margin-left` proportional to its depth - the higher the depth, the wider the margin is. Updates to a job panel, now should happen within the job panel element. If a reflow occurs, it will be a bug, that can be fixed by changes in the CSS or HTML elements.

1 - https://www.sitepoint.com/10-ways-minimize-reflows-improve-performance/

hjoliver commented:
Initial subjective tests in my environment with the complex suite (queue-limited to avoid killing my box with jobs) ... shows this performs pretty damned well :tada:

kinow commented:
Commits squashed/renamed.

kinow commented:
Build passing, just need to check some tests I've disabled and see if they can be adapted for the new code. I've deleted the old tests for `TreeItem` (deleted), and haven't written any for the `InfiniteTree` yet.

The `InfiniteTree` component is a bit trickier to test, as it uses slots, and has all the DOM magic for handling the infinite tree. Might skip writing tests for it for now, wait for manual tests & feedback, then slowly introduce either unit tests or later the e2e tests to cover it.

kinow commented:
This is ready for review.

See comments above for notes about what improved, what got worse. In summary, the component now renders less components simultaneously. However, it spends more CPU time both to run scripts that create the flatten tree for the infinite-tree, as well as adjusting the data in the background when users scroll.

Memory use should be better though. Also, the browser should be less busy updating elements and recalculating styles. This leaves it more time to update the scroll, and render things for the user. The best test is probably run the `master` version, then this branch and compare the overall performance with some suites.

I added some benchmarking results. But note that I did that on my notebook, and not in a reproducible way (i.e. I was basically manually taking screen shots and running examples). So there's room for mistakes, background processes interfering with the tests, etc :+1: 

What I like about the infinite tree is that it gives us more room to manoeuvrer. Before with the recursive tree, given 10_000_000 nodes, I would simply put these 10_000_000 nodes on the DOM, and good luck to the browser to render them.

The infinite tree does increase CPU, but it should get better with incremental updates, and at least we should be able to get these 10_000_00 nodes in a tree, and display only a portion of them each time. Avoiding crashing users' tabs.

oliver-sanders commented on src/components/cylc/tree/InfiniteTree.vue:
```diff
@@ -0,0 +1,202 @@
+      this.nodes = this.tree.nodes
+      this.$nextTick(function () {
+        this.tree.emit('contentDidUpdate')
+      })
```
This emits a `contendDidUpdate` event on the next "tick", do we know the content actually updated? I don't know how these ticks work.

kinow replied:
These ticks are used by Vue to wait for a DOM update.

In the case of the `InfiniteTree`, we have the widget in the DOM. Whenever we manipulate the DOM in Vue, we need to consider whether we are ready to run something, or if we are still with a Virtual DOM node - a VNode.

Later, Vue will translate the VNode to the DOM node. When that happens, Vue gives us a change to execute something - on the next tick.

Not sure if a good explanation, as the weekends gets closer, may English gets worse :smile:

oliver-sanders replied:
Good explanation. So a DOM update will only ever take take one "tick" therefore we can guarentee that the update will be complete after the next "tick".

oliver-sanders commented on src/components/cylc/tree/Tree.vue:
```diff
@@ -0,0 +1,170 @@
+            <span class="mx-1">{{ node.name }}</span>
+          </v-flex>
+        </v-layout>
+        <v-layout class="node-data" :style="nodeStyle(node)" wrap v-else-if="node.__type === 'family'">
```
I don't know if it's possible to define these `<v-layout>` parts separately, would make this template easier to work with.

kinow replied:
Good idea. I will have a look later over the weekend or on Monday to see if it would be something easy to implement.

kinow replied:
Resolving as this was done in #396 . Added a **pending** section to the first/main text of this issue with an item for this.

oliver-sanders commented on src/components/cylc/tree/index.js:
```diff
@@ -59,12 +70,17 @@ function convertGraphQLWorkflowToTree (workflow) {
           name: parent.name,
-          children: []
+          children: [],
+          size: 32,
```
What does the `size` bit do? Should consider making this a constant for easy fiddling.

Is that a visual dimension or to do with the number of nodes in infinite tree?

kinow replied:
Good spot. Deservers some documentation and we can move to a constant I think. That's a required property for the `InfiniteTree` strategy used.

This tells the library what is the size of each element added to the infinite tree. And is later used by the library to calculate how many elements will fit into the screen and calculate the buffer size too.

oliver-sanders replied:
Aah so it's a height estimate in pixels?

kinow replied:
Exactly 👍

oliver-sanders commented on src/styles/cylc/_tree.scss:
```diff
@@ -0,0 +1,35 @@
+    line-height: 1.8em;
+
+    .node-data {
+      margin-left: 12px;
```
Should we use `em` for `margin-left` (and also for `padding-left` in the infinite tree code)?

kinow replied:
Oh, will have to look with more calm why I used `12px`. :+1:

oliver-sanders replied:
It's not a blocker or anything just a quick comment, because the `line-height` is set in `em` above.

kinow commented on src/components/cylc/tree/InfiniteTree.vue:
```diff
@@ -0,0 +1,202 @@
+        return []
+      }
+    },
+    autoOpen: {
```
Before that would be a bit harder, as the collapse/extend state was controlled in the component.

The `InifiniteTree` object keeps state in the data. So we have full control by simply choosing the initial/default state and setting the value in the JavaScript data.

dwsutherland replied:
> If nodes know their depth we could implement "maximum expand depth" in this way.

Nodes do know their depth 😉

kinow replied:
True! But I should have explained better. I think I know one case that doesn't translate directly from the GraphQL results to the Tree view.

I don't remember how we are using depth (i.e. don't remember which nodes in our schema have it). But say we have a workflow with depth=0, then the next step would be cycle points with depth=1 in the tree.

But in our schema, I think the cycle point has a different depth?

We also have special cases, like families. I am not sure if the depth in families is using the correct hierarchy that we need for the tree—which is good I think; our GraphQL schema is independent of component.

Another corner case is the `root` family, which is hidden by default. So the depth has to be corrected for that. So in the end, we need the `depth` matching what we want to display in the Tree component, which is too specific for the component I think?

dwsutherland replied:
> I don't remember how we are using depth (i.e. don't remember which nodes in our schema have it). But say we have a workflow with depth=0, then the next step would be cycle points with depth=1 in the tree.

Depth is defined on all nodes except jobs, and based on Family hierarchy.. Cycle points  in the tree view are really the `root` family in my mind... (BTW - `maxDepth` is a field under `workflow` if it helps with working out minimum margins or something)

kinow commented:
Will start syncing code and rebasing this one in the next days 👍 as well as address pending feedback. Thanks!

oliver-sanders commented:
Thanks, this one was clearly an epic effort :rocket:

kinow commented:
Rebased onto `master`, fixed conflicts. But unfortunately the `Tree` component is now blank. It probably means that the `InfiniteTree` was not loaded correctly. Maybe I need to trigger it after the Lumino widget is created? Will need a few hours/days to investigate.

kinow commented:
Build passing on GitHub. Happy to report that it was easier than expected!

Worked locally on "offline" mode, with dummy data. Also worked with no issues serving two workflows (five & families2). I used the branch of the [pull request that fixes websockets in cylc-uiserver](https://github.com/cylc/cylc-uiserver/pull/107).

Tested on Ubuntu with Chrome & Firefox. No issues.

Also tested on Windows 10Pro with IE Edge. No issues.

It appears to be a bit slower to render the infinite scroll when scrolling very fast, but could be that it was that way before too. It does look very similar to facebook & instagram when you scroll very fast.

Also fixed two possible memory leaks, where data in a map and event listeners were not properly cleaned up.

Ready for review again :+1:

oliver-sanders commented:
> It does look very similar to facebook & instagram when you scroll very fast.

Would be nice to use a [skeleton for scrolling](https://www.altexsoft.com/media/2018/11/skeleton_loading.gif), for future work...

> Ready for review again 

Code looks good.

Unfortunately the tree is rendering really slowly for me with the complex suite - even when held (~5s when I press Page-Down).

Last time I tested it the complex suite was rendering quickly. Are you seeing different performance.

I also got Opera using >2GiB memory which triggered the debugger to hold the session (which may or may not be related to the infinite tree):

![Screenshot from 2020-01-22 11-37-00](https://user-images.githubusercontent.com/16705946/72849554-d1ae0a00-3d0b-11ea-8b72-3fb2698b0419.png)
![Screenshot from 2020-01-22 11-37-52](https://user-images.githubusercontent.com/16705946/72849556-d1ae0a00-3d0b-11ea-96f9-869df386381e.png)

kinow commented:
Thanks for testing it @oliver-sanders ! I may have introduced some bug after rebasing it, or maybe using Lumino with it needs some further work.

>>    It does look very similar to facebook & instagram when you scroll very fast.
>
> Would be nice to use a skeleton for scrolling, for future work...

Good idea! There's a Vuetify component for that too: https://vuetifyjs.com/en/components/skeleton-loaders

hjoliver commented:
Hmmm, I agree - sorry Bruno! - something seems to have gone horribly wrong with the performance on this branch (just tested on the complex suite)

kinow commented:
> 
> 
> Hmmm, I agree - sorry Bruno! - something seems to have gone horribly wrong with the performance on this branch (just tested on the complex suite)

@hjoliver I wonder if accessing the Tree view directly is also not working? From the Workflows Table, there should be a link to view the tree view I think.

hjoliver commented:
> From the Workflows Table, there should be a link to view the tree view I think.

The table rows now link to the lumino tabs view, with the default tree view tab. (I did that last week).

Is there a direct URL I can use to the tree view component (I know there is for the graph view.)

kinow commented:
Sorry, my environment is slightly broken while I finish updating some deps. Just had a look at `src/router/paths.js`, and even though the view `src/views/Tree.vue` exists, it is not bound to any route - meaning we can remove it later, unless it's added back as per #376 

Maybe if you copy the `Graph` entry in `paths.js` and binds it to some route like `/#/tree/:workflowName`  it could work?

kinow commented:
(will finish what I'm doing, then have a look at the Graph component, and then work on this branch again)

hjoliver commented:
> Maybe if you copy the Graph entry in paths.js and binds it to some route like /#/tree/:workflowName it could work?

Thanks for the reminder on that; it does work.  #378

kinow commented:
Nice. If you see performance degradation on that view, then it means it's not Lumino's fault, but some regression that I introduced in some recent-ish commit.

hjoliver commented:
Good news, it is Lumino's fault.  On the complex suite this branch scrolls like a frickin' bat out of hell, sans Lumino.

hjoliver commented:
(Well, I suppose that news is both good and bad, depending on what's going wrong when combined with Lumino!)

kinow commented:
Really good news! I will investigate if the DOM and events integration with Lumino is getting in the way. Maybe I just need to tell Lumino to not bother about whatever is within the widget. 

Thanks!!!! ⭐️

oliver-sanders commented:
Interesting, tested both PRs independently, no problems, you wouldn't have expected the combination to be an issue.

> Maybe I just need to tell Lumino to not bother about whatever is within the widget.

This is a good idea, I didn't actually realise that Lumino was trying to manage it's own portion of the DOM.

kinow commented:
Looking at the performance after rendering a simple tree, with no updates, it's hard to tell any difference in performance. In some runs, actually, the page without Lumino actually loaded faster and spent less time in the scripting stage.

Comparing the performance with a live suite (families2), both also appear to consume the same amount of memory. Took a couple snapshots and inspected the objects in memory, and found nothing too different between both views.

Then, I realized I was using my notebook monitor in one test, and a larger monitor for the other browser window test. So I decided to position both windows on larger monitors... the viewport was still different because of the Lumino BoxPanel.

At that point it occurred to me that the laggyness in rendering the infinite tree, and the performance issues could be related to the viewport. The infinite tree works - apparently - well when it is rendered on an element attached to the normal page.

But when it is attached to a Widget, then the height and width are set by the Lumino widget's proportions. While the infinite tree **might** be calculating the viewport based on the viewport. I tested by making the window viewport smaller (resized the browser window), and then scrolling the widget scroll bar. 

Back to whiteboard, and read the docs & code of the InfiniteTree component. There might be some tip or advice for when you render the infinite tree within a component with the `overflow` CSS property defined, and constrained proportions.

kinow commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cylc-ui",
-  "version": "0.1.0",
+  "version": "0.2.0",
```
I think I added this as I noticed the footer was showing 0.1, but can be moved to a separate PR if necessary.

kinow commented:
Rebased and fixed a few small issues. There were two `<div id="main">`. One that I created, and then another created by Lumino. I think I used the name `main` as I was looking at their examples code in Firefox, with the inspection tool. Didn't realize that element was created dynamically.

Also removed extra events that were not really useful. Reducing the number of events being triggered/listened may result in some performance improvement.

But there is still some issue around how the Tree/InfiniteTree is being added to the Lumino widget. The scroll is not triggering the InfiniteTree to update the elements. That's the reason for the slow response in updating it, especially if compared with the Tree view. So still a bit more work before it's ready for review again.

kinow commented:
How to confirm the Infinite Tree is not working with Lumino:

1. start `NODE_ENV="offline" npm run serve`
2. open two browser windows with `http://localhost:8080/#/workflows/one` (`one` workflow is a static/mocked one)
3. resize one window so that the visible area of the tree is very small
4. compare what happens when you scroll in the resized window, with the larger window

Maximized window:

![image](https://user-images.githubusercontent.com/304786/73399406-0ba29000-434c-11ea-81bd-797961bf9e04.png)

Smaller window (shows only down to the `sleepy` task. After that the Infinite Tree has not loaded.

![image](https://user-images.githubusercontent.com/304786/73399469-2bd24f00-434c-11ea-919e-322cff865b9c.png)

kinow commented:
A few notes about the tree with Lumino (applies to other components too)

- There will be always a difference in performance between a page like `/#/tree/workflowname` and `/#/workflows/workflowname`. When used in a simple view, the component has fewer parents. Probably just the content area `<div>` as a component, and then the rest of the layout. When used with Lumino, it has more `<div>` elements such as `<div id="dock">`, `<div id="main">`, `<div id="main-wrapper">`, and `<div id="workflow-panel">`. Changes to the Tree component, may trigger reflow.
- Lumino widgets and components seem to have more events. Looking at the code generated in Firefox, it's possible to identify the events (marked with a "event" label/tag). Along with the point above, having more event listeners may also degrade the performance.

![image](https://user-images.githubusercontent.com/304786/73400199-87510c80-434d-11ea-8044-e9edf58c7fda.png)

kinow commented:
The `InfiniteTree` is being used mainly to create the flatten data structure. The work of putting the elements in the UI is handled by `vue-virtual-scroller` - that's because `InfiniteTree` will change the DOM, while `vue-virtual-scroller` can use the `InfiniteTree` and create vue components for that.

This is similar to how the `InfiniteTree` React component works (same author).

I'm reading the `vue-virtual-scroller` code today. And noticed it uses [`scrollparent`](https://www.npmjs.com/package/scrollparent) to choose the element that will be bound to the `resize` and `scroll` events - https://github.com/Akryum/vue-virtual-scroller/blob/cec667b948ff7c4c4eb2c41a07c6b860b4c79f80/src/components/RecycleScroller.vue#L518-L522

Using the vue-dev-tools, it looks like it's using the `window` object as target. Meaning that the scroll of the page is triggering the infinite-tree/virtual-scroller to update. And not the scroll of the Lumino widget.

![image](https://user-images.githubusercontent.com/304786/73406889-71971380-435c-11ea-9e01-0b01e5e476fa.png)

Now just need to understand why `scrollparent` didn't pick the parent div, or some other element with `overflow: auto` as the scroll parent target.

kinow commented:
The `scrollparent` JS library simply iterates through each of the given `HTMLElement` parents. Returning either the first to have `overflow`, `overflow-y`, or `overflow-x` with a value that matches "auto" or "scroll".

Put a breakpoint in our code to see what was the `HTMLElement` passed to `scrollparent`.

![image](https://user-images.githubusercontent.com/304786/73407692-c6d42480-435e-11ea-8313-c500fa59817c.png)

The element `vue-recycle-scroller` is a DIV within the Lumino widget. But I was sure we had an `overflow: auto` in one of its parents. Which I confirmed inspecting the code in Firefox.

![image](https://user-images.githubusercontent.com/304786/73407800-1dd9f980-435f-11ea-9a99-ec77b40061d3.png)

My guess now is that I must be creating the Vue component somewhere else and attaching to the Lumino widget too late. By the point it is attached to the widget (or the widget is attached to the BoxPanel?) the parent has already been calculated. It may cause other issues too.

At least that would make sense, and shouldn't be too hard to fix 🤞

kinow commented:
When you add a new Lumino widget with a Tree component, then the Tree component is created by Vue.js in a area that is not displayed. We then cut-paste it into the Lumino widget (document.addElement() I think, which moves a DOM element).

The issue was that the vue-virtual-scroller, when creates the "view" with the visible area and the items to be displayed, tries to locate the parent element with a scroll. When it does, it finds none and binds to the window JS object. Now when you add the widget, we ask the vue-virtual-scroller to reset the listeners. It's a cheap operation that removes/adds listeners, but this time finds the right parent element.

Tested with the complex workflow (used the <=10 items per queue trick, thanks @hjoliver !). I first tested with vue-dev-tools and `npm run build:watch`, and it skyrocketed to ~1GB at one point. Then switched to incognito mode and `npm run build` (production version) and the browser memory kept under 100MB!!! 🎉

kinow commented:
Couple points of the discussion in Riot:

- There is data duplication here (see first comment in this issue, the unchecked item) because the `InfiniteTree` uses `node.state = {...}`, while the GraphQL data contains `node.state = "string"`. Updating one, breaks components use the other.

- The expanded/collapsed state is now owned by the Infinite Tree. That's because the `InfiniteTree` object will use that state, combined with other info such as viewport, element height/widget, etc, to compute how many elements are visible (i.e. if `open=false`, then don't bother adding the children into the calculation!) and pulling these elements from the flat list, creating the DOM elements and updating everything as we scroll (a lot of work, but it still performs surprisingly well).

- Once we have deltas in the UI Server, then we won't have to update the whole tree, and should also be able to control the `node.state.open` when updating each node. As well as handling other cases, such as elements height if necessary.

hjoliver commented:
> The expanded/collapsed state is now owned by the Infinite Tree. 

And the consequence of this is: currently the whole tree reverts to the default collapse/expand state at every update - i.e. user-driven expand/collapse doesn't stick.

kinow commented:
> 
> 
> > The expanded/collapsed state is now owned by the Infinite Tree.
> 
> And the consequence of this is: currently the whole tree reverts to the default collapse/expand state at every update - i.e. user-driven expand/collapse doesn't stick.

Oh, had forgotten to add that part 😬 thanks. Link to the comment explaining it: https://github.com/cylc/cylc-ui/pull/345#issuecomment-568370245 (last paragraph)

kinow commented on src/components/cylc/tree/InfiniteTree.vue:
```diff
@@ -0,0 +1,208 @@
+  mounted () {
+    this.tree = new InfiniteTree({
+      el: this.$refs.tree,
+      ...this.$props,
```
Note to self: this needs to be `this.options.$props`. `this.$props` doesn't seem to work on the `mounted` method (though I think it appears in docs but I think in other methods?)

hjoliver commented:
> try to add the Infinite Tree as a separate component, instead of replacing the current Tree, and use as an extra Lumino widget

(From the Issue description - I think we're not going to bother with this any more - we'll just replace the old tree with the new, as soon as the new one is fully functional ... so re

hjoliver commented:
(Aargh, Close button accident!)

kinow commented:
Thanks for merging the other tree PR. Now I can restart the work on this PR. We should have an updated infinite tree, with no data duplication, next week. Not sure if the expanded,collapsed state will be working without the deltas tho.

kinow commented:
Rebased, but the build should probably fail. I'd added some extra code to fetch cycle points directly, but it needs another change in `GQuery.js`. Will check now whether it's a simple change, or if I should remove that change from this branch.

kinow commented:
Done. Now pending only the deltas, so that the collapse/expand state is not discarded with each update via WebSockets.

kinow commented:
I've reverted #232 in this PR. The reason being that offline mode does not work with that.

The `__typename` property is set by ApolloClient. The offline mode uses `cylc client graphql`, which never sets those values.

It was a good experiment, and now we finally have the answer on whether we can use the `__typename` or not :grimacing:

hjoliver commented:
> I've reverted #232 in this PR. The reason being that offline mode does not work with that.

#232 says "It's not like this property will disappear. So we can drop the __type attribute created in JS, and use the __typename, avoiding duplication."

What was being duplicated (and presumably is again now) - and does that not matter?

kinow commented:
> > I've reverted #232 in this PR. The reason being that offline mode does not work with that.
> 
> #232 says "It's not like this property will disappear. So we can drop the __type attribute created in JS, and use the __typename, avoiding duplication."
> 
> What was being duplicated (and presumably is again now) - and does that not matter?

When the query is retrieved via `ApolloClient` in JS, then we get something like

```js
{
  __typename: 'TaskProxy',
  id: 'foo|id|etc',
  name: 'etc',
  ...
}
```

In the original JS code that I wrote to create the hierarchy, I didn't know about this field, and instead derived a new node (with some duplication :hides:) that looks like:

```js
{
  __typename: 'TaskProxy',
  __type: 'taskproxy',
  id: 'foo|id|etc',
  name: 'etc',
  children: [],
  expanded: false,
  ...
}
```

The change was basically to eliminate the `__typename` and `__type` duplication, and use just the `__typename`.

But it turns out that the `NODE_ENV="offline" npm run serve` uses the code in `src/services/checkpoint` to generate an offline GraphQL response. That is done in a Cylc workflow, that uses `cylc client graphql` to send the GraphQL query.

This does not use `ApolloClient`, and then we don't have `__typename`. We removed `__type`, so the nodes in the offline mode all appear now as normal tree items, without `Task`, `Job`, and other tree specific customization.

So it looks to me that it's not safe to use `__typename` any more, as we require to support cases where we fetch data without `ApolloClient`, which I didn't know before, sorry. That property actually disappeared :)

hjoliver commented:
OK, understood (ish) ... so long as there's no suite data duplication that's fine (I was wondering because #232 was closed by the PR that was supposed to fix that).

kinow commented:
> OK, understood (ish) ... so long as there's no suite data duplication that's fine (I was wondering because #232 was closed by the PR that was supposed to fix that).

Well, that fixed it. But broke the offline mode. I didn't know when I proposed that change that the offline data wouldn't have the `__typename`, i.e. mea culpa :grimacing:

hjoliver commented:
(yes but I meant the suite data elements in the data store .. that's the important stuff)

kinow commented:
Finished re-reading this PR. After the deltas (#458), I think it will be easier to leave this PR open, and use it as reference to add the infinite tree.

The main reason, is that on `master` we have the tree structure in Vuex, and as a `computed` property with the structure the UI needs. But for the infinite tree, we needed hierarchy in an object. So I implemented a tree structure compatible with the infinite tree in this PR.

That's pretty much done in the deltas branch, within the `tree` module and the `CylcTree` object. We just need to:

- add the libraries
- add the components (vue-scroller for instance)
- modify the Tree component using the libs/components
- pass the `CylcTree`'s `root.children` to the `InfiniteTree` constructor
- finally, adjust the `CylcTree` hierarchy if necessary (I tried to add what we needed for the infinite tree in the deltas branch, e.g. node state, but from memory, so I may have forgotten something important)
- test and review :+1: 

If nothing goes wrong, this should be much easier than the deltas PR I think.

kinow commented:
Restarting work here tomorrow morning :+1: should be done by next week :crossed_fingers:

kinow commented:
Not sure if the tests will pass.

Rebased, had issues with dependencies that changed, so had to downgrade/upgrade some dependencies to get everything working together.

Also had another issue since now we have the Vuetify skeleton loader component in between the `Tree` component and the actual tree. So it's one more element in the hierarchy, and that loads the `Tree` dynamically (this `Tree` component contains the infinite tree).

I had some code in `mounted()` calling a method to reset listeners to avoid issues when resizing windows, that was failing because of the skeleton loader.

Tested with `five`, `families2`, and `complex`. Memory was good, even though I was using the development version with vue-dev-tools.

Only issue found was that the job-leaf node was not expanding. I probably forgot to include it, or messed up when merging. We also changed `__typename` or `__type` to `type`. So I may have to adjust that. But it's working OK, deltas + infinite tree. Just needs a few kicks here and there, and then should be ready for review again :+1:

kinow commented:
## Benchmarking

Performance is still looking good.

Started the `complex` workflow, and left it running for 15 minutes, occasionally scrolling to see if the data was populating the tree correctly.

**Before** the workflow was loaded in the UI, my tab had **12 MB** used. It went up to **26 MB**, and eventually up to **32 MB**.

The complex workflow keeps this memory usage for a while (noticed from previous runs) until it starts to use a little more. In this case, it **varied between 30 and 39 MB**. It did not reach 40 MB.

This is the performance of the complex workflow for **deltas**.

![image](https://user-images.githubusercontent.com/304786/82172508-74733f80-991e-11ea-843e-a45bbcc5cb3c.png)

I got similar output for the **deltas with infinite tree**. Main difference being that it spent more time in the scripting part. I believe that's due to the infinite tree having to re-calculate what is displayed or not, and flattening the tree.

However, if you observe the Rendering and Painting, it should be clear that the browser is less busy with these two tasks.

If you have a workflow that doesn't update every few seconds (maybe every minute or so) the scripting won't be an issue. I didn't notice any issue for my browser. I could scroll the tree view, and navigate with no issues.

The main advantage now is that when you scroll for a very large workflow, it wont' suddently have to start scripting + rendering + painting, and increasing the amount of memory used. Instead, scripting should increase a bit, but the rest should remain similar, as we will only render the N tree items displayed (plus some bufer).

Also tried scrolling very fast to see if I would notice a lag (as in GitHub actions) but the performance was quite good. That could be a problem for users with slow network, as my servers are all running locally.

![image](https://user-images.githubusercontent.com/304786/89359649-71683e00-d71a-11ea-943a-d867e12463ca.png)

kinow commented:
I found one issue while testing the complex workflow, with jobs and the job details.

![image](https://user-images.githubusercontent.com/304786/89359933-184cda00-d71b-11ea-80ed-9b0ef00c2c61.png)

I think I might be calling a method of a wrong object somewhere, or maybe referencing the wrong object in a method parameter.

Performance is OK. Tests are OK too. So one issue before it can be properly reviewed :+1:

kinow commented:
The issue appears to be that the Infinite Tree object, isn't defining the `depth` of some jobs correctly. Not sure why.

![image](https://user-images.githubusercontent.com/304786/89361605-cd819100-d71f-11ea-9965-21762bc6d737.png)

The only node with `depth: 1` must be the workflow I think. Everything else has `depth: 2` or greater.

kinow commented:
Note: the depth is not set in the `infinite-tree` library, but in the `flattree` (same author). Had a look at the code, and couldn't understand why it wouldn't compute the correct `depth`.

There are some jobs that are being created with `depth: NaN` :thinking:

hjoliver commented:
> I tried setting the depth when creating the nodes... but with deltas, we may receive a family proxy that is deeply nested, without knowing its depth (i.e. we could have several parent family proxies that are in other parts of the delta message, yet to be processed). 

If I understand what "depth" means in this context, the family tree structure (including depth) of a workflow is static so we could potentially compute that server-side and make it an attribute of tasks and families - if that would help?

kinow commented:
> > I tried setting the depth when creating the nodes... but with deltas, we may receive a family proxy that is deeply nested, without knowing its depth (i.e. we could have several parent family proxies that are in other parts of the delta message, yet to be processed).
> 
> If I understand what "depth" means in this context, the family tree structure (including depth) of a workflow is static so we could potentially compute that server-side and make it an attribute of tasks and families - if that would help?

That would definitely help, though I am not sure if we know what is the depth of each node? If we did, that would improve the UI scripting time.

kinow commented:
Found another interesting thing, I added a `console.log` in the `createJobNode`. That's called for each Job object returned in the `deltas.added`. Apparently, I'm getting it duplicated sometimes.

![image](https://user-images.githubusercontent.com/304786/89369841-5655f800-d733-11ea-8d9b-f5936ca39ff6.png)

Could be related, as I'm only having issues with jobs and job details nodes.

hjoliver commented:
> though I am not sure if we know what is the depth of each node? 

Do you just mean the tree depth?  We may not compute that now, but it would be pretty easy. E.g. in the old "family" example, this structure never changes:
![shot](https://user-images.githubusercontent.com/2362137/89370181-3a9f2180-d734-11ea-94bb-dd6769f38b10.png)

kinow commented:
> E.g. in the old "family" example, this structure never changes:

Ah! Of course, so it wouldn't be so expensive to send that information from the backend. I will need to confirm that the `flattree` JS library won't try to compute it again, but for the future I think that'd be nice if we added this info in the data sent from the backend (we can fork the `flattree` library if necessary, or give flat data to the infinite tree, there's an issue for that I believe).

kinow commented:
Ha! Eureka!!! Nothing like one day doing nothing but house chores to take your mind out of a problem like this one.

Turned the laptop on to read a book and suddenly some part of my brain clicked!

What I was seeing in the infinite tree is quite simple. It is a job node, that is parent-less (i.e. no `firstParent`). The hierarchy is correct, but probably after an update delta told the tree what is the missing parent.

In other words, this is the known issue that we have with the data store (before SoD). So when reviewing this pull request, you won't see that error any more.

Instead of a browser console error, the node is rendered by the infinite tree, but in a buggy state, without its parent. When you expand the `TaskProxy`, it will display the child node, which has `depth: 0` I think, which is not incorrect.

The infinite tree does not calculate the depth, it only displays the nodes one after the other. Our Vue code adds a left-padding to help viewing the hierarchy. Who defines the `depth` is the `flattree` module. When the data is given to `flattree`, it uses the child-parent relationship to define the `depth`.

It does not complain about the missing parent, and instead resets the depth.

In other words, ready for review. If you see a `Job` behaving a bit odd, that's the issue with the missing parent. Really glad it happened before, during the work on deltas. Otherwise I would spend a loooong time to debug it, as issues like this are much harder to detect with the infinite tree (you don't have the whole tree displayed, so VueDevTools is also limited, and there are extra structures like the `InfiniteTree` object, its `nodeTables` (a `LookupTable` object), and the `InfiniteTree.nodes`.

kinow commented:
Found a few more things that changed since this PR were created, so removed unused code, and simplified where possible. :+1:

kinow commented on src/components/cylc/tree/Tree.vue:
```diff
@@ -16,118 +16,178 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * Offset used to move nodes to the right or left, to represent the nodes hierarchy.
+ * @type {number} integer
+ */
+const NODE_DEPTH_OFFSET = 30
```
Yup, but this value needs to be a `number`, as it's used to calculate the `padding-left` and `margin-left` values further down. e.g.

```js
    getNodeStyle (node) {
      return {
        'padding-left': `${node.state.depth * NODE_DEPTH_OFFSET}px`
      }
    },
```

oliver-sanders replied:
Ah sorry, I just meant could you add a the word "pixels" into the comment somewhere so we know what the units are meant to be.

kinow commented on src/components/cylc/tree/Tree.vue:
```diff
@@ -16,118 +16,178 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+            v-if="!['job', 'job-details'].includes(node.type)"
+            shrink
+          >
+            <span v-html="togglerContent(node)"></span>
```
Good idea! I've replaced it by CSS here: https://github.com/cylc/cylc-ui/pull/345/commits/640487d2a609d855a23b86e5e6515ec9d64251c1

Not sure if the best approach, but was the first idea I had.

kinow commented on src/styles/cylc/_tree.scss:
```diff
@@ -0,0 +1,90 @@
+  .type {
+    margin-right: 10px;
+  }
+  $arrow-size: 15px;
```
Good catch! Removing it now :+1:

kinow commented on src/components/cylc/tree/InfiniteTree.vue:
```diff
@@ -0,0 +1,194 @@
+// TODO: import InifiniteTree from 'infinite-tree' is not working with dependency installed from git commit. Replace once there is a new release of infinite-tree
+import InfiniteTree from '../../../../node_modules/infinite-tree/dist/infinite-tree'
+
+const lcfirst = (str) => {
+  str += ''
+  return str.charAt(0).toLowerCase() + str.substr(1)
+}
```
This was copied from the example React component written by the author of the `InfiniteTree` component.

https://github.com/cheton/react-infinite-tree/blob/e9258998254d8d0e98a1640b87d1d13f6237cffc/src/InfiniteTree.jsx#L6

It's used only within this file (not exported), to capitalize the event handler names, 

````
>lcfirst('onContentWillUpdate'.substr(2))
contentWillUpdate
```

Luckily the events declaration is similar in React and Vue. I copied most of what I could from his React code for the initial component, only replacing the React `VirtualList` by Vue `RecycleScroller`. But we can improve it if you think it's necessary.

kinow commented on package.json:
```diff
@@ -22,43 +22,43 @@
     "graphiql": "^0.17.5",
     "graphql": "^14.5.8",
     "graphql-tag": "^2.10.1",
+    "infinite-tree": "git+https://git@github.com/cheton/infinite-tree.git#7bf037ee79b8aa702f5ee605b8ec218bab2533b4",
```
Good question. I thought I had left a comment somewhere... I clearly remember adding a comment above this line, but I had thought I was writing YAML and of course `yarn install` failed.

I had to change the GitHub action linter to allow github.com too, but I only mention the issue in the comment there. Let me chase and see what was the issue.

kinow replied:
Ah, here it's.

I've downgraded to the latest 1.16.2 release from NPM.

![image](https://user-images.githubusercontent.com/304786/89841513-30af6f80-dbc7-11ea-9b11-7ecd5762fbbf.png)

I recall now spending a long time staring at the screen, not understanding why it wouldn't work. After trying every possible setting, and reading the library code, I decided to check if there were un-released commits.

As there were some, I pointed to the latest commit, and everything worked. I cannot tell which commit fixes it, but the next release should include it. And I haven't found any other issues using this commit, so far.

Should we create a follow-up issue set for 0.3, saying why we have that pinned commit in the `package.json`, and that prior to a release of Cylc UI, we should check if there's a version newer than 1.16.2, and if so, what steps we must take (e.g. update version in package.json, remove github.com entry from linter)?

kinow commented on src/components/cylc/tree/index.js:
```diff
@@ -18,6 +18,47 @@
+]
+
+/**
+ * The height of each tree item. In the infinite tree, it is used as value for the `.size` property.
```
>Another quick question, what happens if the user changes the font-size?

Huh. Hadn't thought about that. I did a quick test, first with the workflow held. And everything worked fine.

The I released it, and resized window so that there were two scroll bars (one for the window area, one for the tree view). And started scrolling both, while increasing/decreasing the font size. Apparently it worked OK.

We have an issue for accessibility, so I think it would still be a good idea to leave a comment there saying that the infinite tree calculates heights with pixels (for now, will check if we can use `em` instead) and that we need to make sure changing the font size doesn't cause unexpected issues in the tree view.

WDYT?

kinow replied:
>A potential enhancement for a later PR, could we write a functional test which checks that the height is in-fact 32 pixels so that any theme changes don't accidentally break the infinite tree.

+1!

And this would be a good issue for any new contributor.

kinow commented on src/styles/cylc/_tree.scss:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    .node-data {
+      margin-left: 6px;
```
Good idea. I think so far both `Workflow` and `Tree` components have accumulated quite a bit of `px` definitions (sorry!). Will need just some time to investigate whether it's something easy to change here now, or as a follow-up PR, probably fixing in all components where possible.

kinow commented on package.json:
```diff
@@ -22,43 +22,43 @@
-    "cytoscape-undo-redo": "^1.3.1",
+    "cytoscape-navigator": "^1.3.3",
+    "cytoscape-popper": "^1.0.5",
+    "cytoscape-undo-redo": "^1.3.2",
```
Mea culpa. I upgraded all dependencies initially, as the InfiniteTree wasn't working correctly. Later realized the problem was a commit that wasn't released yet, but forgot to remove some dependencies that were upgraded.

kinow commented:
Thanks for the review @oliver-sanders ! Addressed comments. Still pending the `px`->`em`/`rem`, will take a look to see how hard it would be to use relative dimensions. Only concern is whether it will cause any issues with how the `InfiniteTree` object calculates the height of the tree entries.

>I'm seeing an issue switching between workflows where the tree never loads (see screen grab), let me know if I can provide any useful debug info.

It would be good to know if that happens on `master` too? If so, I suspect it would be due or related to #454. Otherwise more likely to be an issue with some change in here. Could you try to replicate it with `master`, please?

kinow commented on src/components/cylc/tree/InfiniteTree.vue:
```diff
@@ -0,0 +1,194 @@
+    },
+    rowHeight: {
+      type: Number,
+      default: 32
```
It was from the React example. This, and the `tabIndex` above, are not actually used.

They are not for the `InfiniteTree`, but for the component that wraps the infinite tree recycling objects. The React example uses both, `tabIndex` and `rowHeight`, but the `RecycleScroller` didn't need it.

I used it initially, but that broke with the job details, as it has something like `7 * 32` height, maybe more or less... so instead the `RecycleScroller` has no `item-size`. This way, it spends a bit more of scripting time calculating the height of each element dynamically.

A price we need to pay as not all the elements are sized equally in our tree.

kinow replied:
Removed both entries!

kinow commented:
I've actually reproduced the issue @oliver-sanders . It is some buggy logic in how I'm handling the existing subscriptions when the user switches views. Shouldn't be too hard to fix (:crossed_fingers: ).

kinow commented:
On `em`/`rem` instead of `px`, that's not possible due to how the `vue-virtual-scroller` creates the virtual elements in the visible area (the infinite tree is just providing the flat tree structure).

It uses the value to create a `translateY` rule in each tree item.

![image](https://user-images.githubusercontent.com/304786/90078348-b7e01d00-dd58-11ea-9c3a-07aa5e77be3c.png)

It appears to be working OK when I change the browser zoom and viewport area.

kinow commented:
Rebased

oliver-sanders commented:
> On em/rem instead of px, that's not possible due to how the vue-virtual-scroller creates the virtual elements in the visible area

Fair enough, makes sense.

> It appears to be working OK when I change the browser zoom and viewport area.

Ah so it does, and the style seems to hold together perfectly.

So long as accessibility is fine I'm happy, looks good to me:

<img width="500" alt="Screenshot 2020-08-13 at 10 55 20" src="https://user-images.githubusercontent.com/16705946/90121495-f1faf000-dd53-11ea-8e53-d7ed6964c0a3.png">


<img width="500" alt="Screenshot 2020-08-13 at 10 55 00" src="https://user-images.githubusercontent.com/16705946/90121480-ef989600-dd53-11ea-830a-525a428056c6.png">


**Sorry, but I've found another bug :(**

If I expand items in the tree, then scroll up or down, sometimes items get cut off of the front or end of the tree. I can replicate this with the "one" suite in the mocked data:

<img width="479" alt="Screenshot 2020-08-13 at 10 57 36" src="https://user-images.githubusercontent.com/16705946/90121657-253d7f00-dd54-11ea-9b17-6a0b77983b27.png">

<img width="479" alt="Screenshot 2020-08-13 at 10 57 09" src="https://user-images.githubusercontent.com/16705946/90121654-24a4e880-dd54-11ea-8f7a-ab6342eee6cb.png">

Expand/collapse or switching tab seems to be enough to un-stick the tree.

I guess this sort of thing is always a risk with a virtual tree so it might be very difficult to fix this entirely. If we could reduce the likelyhood, and make it clearer to the user that there is content that they can't see (e.g. with a skeleton for the missing items) that would be enough.

hjoliver commented:
The expand/collapse triangles and in-line job icons aren't independent between lumino tabs (is this a known issue?). Here I expanded `cfoo` in the tab on the right:

![Screenshot from 2020-08-13 23-16-29](https://user-images.githubusercontent.com/2362137/90129676-2e166b00-ddbd-11ea-8db9-55e49e397eb6.png)

oliver-sanders commented:
> The expand/collapse triangles and in-line job icons aren't independent between lumino tabs

Ah, dang good spot, I guess the expand/collapse state is stored in the shared tree rather than locally in the view instance.

hjoliver commented:
> If I expand items in the tree, then scroll up or down, sometimes items get cut off of the front or end of the tree. 

I've played around quite extensively with the complex suite and a smaller suite. So far I've seen this **all the time** in a second tab (if I have a second tab open) but never in the first (or only, if no second one is open) tab.

If I have a second tab open, the view never updates on scrolling.  If I scroll the most of the rendered items off-screen, then expand or collapse one of the remaining nodes, at that point the whole tab updates.

hjoliver commented:
(Scrolling performance seems really good :+1: )

kinow commented:
I tested with one and multiple workflows, and with tree, and tree+graph. But never tested tree+tree :grin: good catch @oliver-sanders ! And @hjoliver is right about the expand/collapse not being independent.

There is a single tree, backing both tree components. When you expand one, it emits an event telling the infinite tree that we want to expand a node.

The node's state is changed, reflecting the new `state: { open : true }` modification. And it emits some other events and performs a few more actions, which I am currently digging into the code to understand more of how it works.

Both issues reported here are due to this state-sharing issue. When you expand on one tree component, the other component receives the new state, and changes the expand/collapse icon. However, the elements are not rendered.

So if you have a limited viewport, with scroll active, and expand one one component, the other will change its state without rendering the elements. When you scroll on the other window, the component will think it needs to show the children elements of the expanded node, but as these elements were not rendered, you see only a blank area.

![image](https://user-images.githubusercontent.com/304786/90198296-68b3ee00-de25-11ea-801f-098c17bb03c9.png)

I believe if we are able to tell both components to change the state, and render the newly expanded nodes, it should work correctly. Then we can discuss having separate states per widget... or maybe having a toggle button to link the widgets.

kinow commented:
Actually there's one more issue. Related to the double scroll bars. If you scroll up or down, and you have the view area scroll, and also the component scroll, both active. Then looks like the outer scroll bar needs to emit an event to tell the tree we need to refresh it.

Without that happening, you can end up with empty areas in the tree, due to elements not being rendered.

![image](https://user-images.githubusercontent.com/304786/90198670-6d2cd680-de26-11ea-847f-224df22fe22b.png)

![image](https://user-images.githubusercontent.com/304786/90198701-8df52c00-de26-11ea-88d8-07f845036dd3.png)

We already have some code taking care of events when you add other widgets, to tell our tree we want it to update. It might be just a matter of finding what else we need to listen to and tell the tree to update.

p.s. the tree component uses vue-virtual-scroller, which uses [the intersection observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API), which has a few issues with browsers. It could be one of these issues too :nerd_face:

kinow commented:
## Scrollbar issue

Here's how I'm troubleshooting this first issue. First I resized the browser view port area so that I had both scroll bars visible, and successfully reproduced the issue.

Then I used VueDevTools to find the `RecycleScroller` component, and stored a reference as `$vm0`. Then started testing the methods of that component instance, to see what would fix the issue.

- `$vm0.handleResize()` fixes it
- `$vm0.handleScroll()` fixes it
- `$vm0.updateVisibleItems()` fixes it

These are the three methods that I found in the object that fixed it.

WIP

kinow commented:
## shared state (expand/collapse) issue

The `CylcTree` object, that is created for and with the deltas, stores in each node created an attribute with the state for the collapsed/expanded featured.

```
tree-entry: {
  id: 'unique-id',
  size: 32,
  node: notReplicatedNodeFromGraphQL,
  state: {
    open: !['job', 'job-details'].includes(node.type) // this is kind of what we are doing in the code, expanding everything but jobs and leaf nodes
  }
}
```

If we have two or more Lumino widgets added to the Workflow view, each widget will have a `TreeWrapper`. And each wrapper has an `InfiniteTree`, but the `InfiniteTree`s are created using the same `CylcTree`, without duplicating the data.

When you expand or collapse a node, we call the `InfiniteTree.openNode(node)` or `tree.closeNode(node)`. That changes the state of the node and tells the tree it needs to recompute what is displayed or not.

If you have two infinite trees, `A` and `B`, and you expand a node in `A`, the node will be expanded in both `A` and `B`.

**But only `A.openNode(node)` is called. So the tree `B` doesn't understand it needs to update.** I believe that's causing the issue at the moment.

I think we have a few possible solutions so far:

1. call `openNode` for each tree in the `Workflow.vue` component, i.e. iterate the widgets, if type of widget is tree, call `$component.tree.openNode(node)`.
2. move the state out of the tree, and into the component, so each widget controls the state separately
3. check if the infinite tree and/or vue-recycle-scroller support the state in some other way (a function perhaps, that is applied only to the local instance of the flattree)
4. explore hacking the flattree object, injecting the state or intercepting the change of state if coming from a different tree?

kinow commented:
Closed as per Riot message :tada: 

>tree view decision
>
With agreement from the UK end, via oliver sanders , we are going to shelve the infinite tree for now.
>
>We may come back to virtual scrolling in future if we need it for large workflows, but perhaps not with a collapsing tree view. For the moment, delta-updates seem to have nicely revitalized the simpler plain tree view, and low-n windows will make it even better again soon. My guess is virtual scrolling works well for an ever-extending flat list of items, but our data (particularly in collapsing tree view) is not so simple.
